faunadb-java/src/main/java/com/faunadb/client/types/Value.java

```diff
@@ -6,335 +6,488 @@
+   *
+   * @return the boolean value of this node.
+   * @throws ClassCastException if can not coerced to {@link LocalDate}.
    */
   LocalDate asDate();
 
   /**
-   * Coerces this node into an ordered list of nodes.
-   * @return an ordered list of response nodes, or null.
+   * Attempts to coerce this node into a {@link LocalDate}.
+   *
+   * @return an {@link Optional} type with the coerced value.
+   */
+  Optional<LocalDate> asDateOption();
+
+  /**
+   * Coerces this node into a {@link ImmutableList} of nodes.
+   *
+   * @return a immutable list of nodes.
+   * @throws ClassCastException if can not coerced to {@link ImmutableList}.
    */
   ImmutableList<Value> asArray();
 
   /**
-   * Coerces this node into a dictionary of nodes.
-   * @return a dictionary of nodes, or null.
+   * Attempts to coerce this node into a {@link ImmutableList} of nodes.
+   *
+   * @return an {@link Optional} type with the coerced value.
+   */
+  Optional<ImmutableList<Value>> asArrayOption();
+
+  /**
+   * Coerces this node into a {@link ImmutableMap} of nodes.
+   *
+   * @return a immutable map of nodes.
+   * @throws ClassCastException if can not coerced to {@link ImmutableMap}.
    */
   ImmutableMap<String, Value> asObject();
 
   /**
+   * Attempts to coerce this node into a {@link ImmutableMap} of nodes.
+   *
+   * @return an {@link Optional} type with the coerced value.
+   */
+  Optional<ImmutableMap<String, Value>> asObjectOption();
+
+  /**
    * Coerces this node into a {@link Ref}.
-   * @return a Ref, or null.
+   *
+   * @return a immutable map of nodes.
+   * @throws ClassCastException if can not coerced to {@link Ref}.
    */
   Ref asRef();
 
   /**
-   * Coerces this node into a {@link Set}.
-   * @return a Set, or null.
+   * Attempts to coerce this node into a {@link Ref}.
+   *
+   * @return an {@link Optional} type with the coerced value.
    */
-  Set asSet();
+  Optional<Ref> asRefOption();
 
   /**
-   * Coerces this node into a {@link Page}.
-   * @return a Page, or null.
+   * Coerces this node into a {@link SetRef}.
+   *
+   * @return a immutable map of nodes.
+   * @throws ClassCastException if can not coerced to {@link SetRef}.
    */
-  Page asPage();
+  SetRef asSetRef();
 
   /**
-   * Coerces this node into an {@link Instance}.
-   * @return an Instance, or null.
+   * Attempts to coerce this node into a {@link SetRef}.
+   *
+   * @return an {@link Optional} type with the coerced value.
    */
-  Instance asInstance();
+  Optional<SetRef> asSetRefOption();
 
   /**
-   * Coerces this node into a {@link Key}.
-   * @return a Key, or null.
+   * Extract a specific field from this node.
+   * <p>
+   * <b>Example:</b>
+   * <pre>{@code
+   * // node = { "name": "jhon" }
+   * node.get("name").asString() // "jhon"
+   * }</pre>
+   *
+   * @return the value under the key.
+   * @throws IllegalArgumentException if the field does not exists.
    */
-  Key asKey();
+  Value get(String key);
```

Comment:
This is where I'm heading to:

```java
Field<String> stringFiel = Field.at("data", "testField", "string").asString();
System.out.println(instance.get(stringFiel)); //sup
```

I think it is useful for the user to be able to extract the fields into constants. Although, it is more verbose for simple queries.

Would you want to keep another helper method like:

```java
instance.get("data", "testField", "string").asString();
```

It could internally use the `Field` class navigation anyway...